### PR TITLE
Apply the latest artifactId in pluginMarkerMaven POM.

### LINKED ISSUE #23551: Invalid artifactId is used in pluginMarkerMaven POM
### Context
I have a certain project structure on the file system, this does not match one to one with artifact names generated in `mavenCentral()`. I think this is quite usual setup. I recently wanted to enable marker publication, but the markers are not working, because their POM contains invalid artifact IDs which point to the directory structure rather than the actual artifact ID under which the project is published. Example below will make this very clear.

### Steps to Reproduce
1. Clone https://github.com/TWiStErRob/repros/tree/master/gradle/gradlePluginMarker-earlyArtifactId
2. `gradlew publishToMavenLocal`  
   (or more specifically `gradlew generatePomFileForPluginMavenPublication generatePomFileForMyPluginPluginMarkerMavenPublication`)

<details><summary>Observe the output POM files</summary>

Observe these two files:
 * `published/build/publications/pluginMaven/pom-default.xml`
 * `published/build/publications/myPluginPluginMarkerMaven/pom-default.xml`

Or alternatively the published artifacts in your `~/.m2` (`mavenLocal()`) folder:
 * `~/.m2/com/example/different-name/1.2.3/different-name-1.2.3.pom`
 * `~/.m2/com/example/my-plugin/com.example.my-plugin.gradle.plugin/1.2.3/com.example.my-plugin.gradle.plugin-1.2.3.pom`

Note: the name and location of these files are all correct based on the setup, which is in short:
```kotlin
group = "com.example"
version = "1.2.3"
afterEvaluate {
    publishing.publications.named<MavenPublication>("pluginMaven").configure { artifactId = "different-name" }
}
gradlePlugin.plugins.create("myPlugin") { id = "com.example.my-plugin" }
```
</details>

### Current Behavior
This is the marker POM.
The referenced artifact doesn't exists, it assumes that the folder name is the same as the published artifact ID.
```xml
<project>
  <groupId>com.example.my-plugin</groupId>
  <artifactId>com.example.my-plugin.gradle.plugin</artifactId>
  <version>1.2.3</version>
  <packaging>pom</packaging>
  <dependencies>
    <dependency>
      <groupId>com.example</groupId>
      <artifactId>published</artifactId> <!-- ................... PROBLEM HERE -->
      <version>1.2.3</version>
    </dependency>
  </dependencies>
</project>
```

### Expected Behavior
```xml
<project>
  <groupId>com.example.my-plugin</groupId>
  <artifactId>com.example.my-plugin.gradle.plugin</artifactId>
  <version>1.2.3</version>
  <packaging>pom</packaging>
  <dependencies>
    <dependency>
      <groupId>com.example</groupId>
      <artifactId>different-name</artifactId> <!-- ................... CORRECT HERE -->
      <version>1.2.3</version>
    </dependency>
  </dependencies>
</project>
```

### Your Environment

Build scan URL: don't think it's relevant given the details below

### Further details
I actually found the problem here:
`org.gradle.plugin.devel.plugins.MavenPluginPublishPlugin#createMavenMarkerPublication`
 * This method reads the coordinates into local variables `pluginGroupId`, `pluginArtifactId` and `pluginVersion`.
 * Then afterEvaluate in my build.gradle executes
 * Then `generatePomFileFor...Publication` task executes the `getPom().withXml` block
 * Which reads the cached local variables

I'm forced to use `afterEvaluate` in build.gradle because of the `afterEvaluate` in `MavenPluginPublishPlugin#apply`.

There's literally no way to correctly configure this whether using the `plugins` block or `apply(plugin =`.

Bar rewriting the two plugins with full `Property<T>` support, the solution would be to capture the `coordinates` variable in the `Action<XmlProvider>` block and read the group/artifact/version lazily during task execution.

### Workaround
The only thing I can do is to either re-do the work of `java-gradle-plugin` (`MavenPluginPublishPlugin`) by registering the publication myself or by generating the marker myself. Neither of these is acceptable given the complexity and all the laziness work that has been done in Gradle over the last years to solve exactly these kinds of problems.

### PR DESCRIPTION
**Pull request by Chaoba**
Signed-off-by: Yanshun Li <qqliys@126.com>

Fixes #23551

### Context

While generating the pom file for maven maker publication,  it will apply the cached values of`pluginGroupId,  pluginArtifactId,  pluginVersion`. However, these values could be changed before executing the generating task.
In this pr, instead of applying the cached values, the latest values will be applied to fix this issue.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation

### CODE REVIEW

**Review by eirnym — APPROVED** [review]

File: subprojects/plugin-development/src/integTest/groovy/org/gradle/plugin/devel/plugins/JavaGradlePluginPluginPublishingIntegrationTest.groovy [hunk 1]
```diff
@@ -170,6 +171,40 @@ class JavaGradlePluginPluginPublishingIntegrationTest extends AbstractIntegratio
...
+        publishToMaven()
+
+        and:
+        buildFile << """
+            publishing {
+                afterEvaluate {
+                    publications {
+                        getByName("pluginMaven"){
+                            configure{
+                                artifactId = "foo-new"
```
Comment (TWiStErRob) [inline subprojects/plugin-development/src/integTest/groovy/org/gradle/plugin/devel/plugins/JavaGradlePluginPluginPublishingIntegrationTest.groovy:188]: To increase coverage, I would change all 3 things that are affected in the test.

Comment (Chaoba) [reply subprojects/plugin-development/src/integTest/groovy/org/gradle/plugin/devel/plugins/JavaGradlePluginPluginPublishingIntegrationTest.groovy:188]: Sounds good. Test case updated.

**Review by TWiStErRob — APPROVED** [review]

**Review by ov7a — APPROVED** [review]

**Comment by ov7a** [comment]
@bot-gradle test and merge

**Comment by ov7a** [comment]
@bot-gradle test and merge

**Comment by TWiStErRob** [comment]
@ov7a according to https://github.com/gradle/gradle/blob/615f18ad22c40cb62181c6bff8fc1f69ef375c2b/released-versions.json#L3 this will be in 8.1 RC not 8.2 RC.

**Comment by ljacomet** [comment]
We branched on Monday for 8.1. This will be in 8.2 indeed.
The file to check is `version.txt` at the root of the project. It was [updated](https://github.com/gradle/gradle/commit/b05c9fb3b2fe248d76c9c2a21300e0c06188bde1) recently.

**Comment by TWiStErRob** [comment]
Ah, thanks!

**Comment by eirnym** [comment]
Thank you

**Pull request merged**